File: c/zqfield.cpp
#include "zqfield.h"

ZqField::ZqField(PBigInt ap) {
    mpz_init2(tmp, 1024);
    mpz_init_set(p, *ap);
    mpz_init_set_ui(zero, 0);
    mpz_init_set_ui(one, 1);
}

ZqField::~ZqField() {
    mpz_clear(tmp);
    mpz_clear(p);
    mpz_clear(zero);
    mpz_clear(one);
}

void ZqField::add(PBigInt r, PBigInt a, PBigInt b) {
    mpz_add(*r,*a,*b);
    if (mpz_cmp(*r, p) >= 0) {
        mpz_sub(*r, *r, p);
    }
}

void ZqField::sub(PBigInt r, PBigInt a, PBigInt b) {
    if (mpz_cmp(*a, *b) >= 0) {
        mpz_sub(*r, *a, *b);
    } else {
        mpz_sub(tmp, *b, *a);
        mpz_sub(*r, p, tmp);
    }
}

void ZqField::mul(PBigInt r, PBigInt a, PBigInt b) {
    mpz_mul(tmp,*a,*b);
    mpz_fdiv_r(*r, tmp, p);
}

void ZqField::div(PBigInt r, PBigInt a, PBigInt b) {
    mpz_invert(tmp, *b, p);
    mpz_mul(tmp,*a,tmp);
    mpz_fdiv_r(*r, tmp, p);
}

void ZqField::idiv(PBigInt r, PBigInt a, PBigInt b) {
    mpz_fdiv_q(*r, *a, *b);
}

void ZqField::mod(PBigInt r, PBigInt a, PBigInt b) {
    mpz_fdiv_r(*r, *a, *b);
}

void ZqField::lt(PBigInt r, PBigInt a, PBigInt b) {
    int c = mpz_cmp(*a, *b);
    if (c<0) {
        mpz_set(*r, one);
    } else {
        mpz_set(*r, zero);
    }
}

void ZqField::eq(PBigInt r, PBigInt a, PBigInt b) {
    int c = mpz_cmp(*a, *b);
    if (c==0) {
        mpz_set(*r, one);
    } else {
        mpz_set(*r, zero);
    }
}

void ZqField::gt(PBigInt r, PBigInt a, PBigInt b) {
    int c = mpz_cmp(*a, *b);
    if (c>0) {
        mpz_set(*r, one);
    } else {
        mpz_set(*r, zero);
    }
}

void ZqField::leq(PBigInt r, PBigInt a, PBigInt b) {
    int c = mpz_cmp(*a, *b);
    if (c<=0) {
        mpz_set(*r, one);
    } else {
        mpz_set(*r, zero);
    }
}

void ZqField::geq(PBigInt r, PBigInt a, PBigInt b) {
    int c = mpz_cmp(*a, *b);
    if (c>=0) {
        mpz_set(*r, one);
    } else {
        mpz_set(*r, zero);
    }
}

void ZqField::neq(PBigInt r, PBigInt a, PBigInt b) {
    int c = mpz_cmp(*a, *b);
    if (c!=0) {
        mpz_set(*r, one);
    } else {
        mpz_set(*r, zero);
    }
}

int ZqField::isTrue(PBigInt a) {
    return mpz_sgn(*a);
}

void ZqField::copyn(PBigInt a, PBigInt b, int n) {
    for (int i=0;i<n; i++) mpz_set(a[i], b[i]);
}

File: src/zqfield.js
const bigInt = require("big-integer");
const assert = require("assert");

module.exports = class ZqField {
    constructor(p) {
        this.p = p;
        this.bitLength = p.bitLength();
        this.mask = bigInt.one.shiftLeft(this.bitLength - 1).minus(bigInt.one);
    }

    add(a, b) {
        let res = a.add(b);
        if (res.geq(this.p)) {
            res = res.minsu(this.p);
        }
        return res;
    }

    sub(a, b) {
        if (a.geq(b)) {
            return a.minus(b);
        } else {
            return this.p.minus(b.minus(a));
        }
    }

    neg(a) {
        if (a.isZero()) return a;
        return this.p.minus(a);
    }

    mul(a, b) {
        return a.mul(b).mod(this.p);
    }

    lt(a, b) {
        return a.lt(b) ? bigInt(1) : bigInt(0);
    }

    eq(a, b) {
        return a.eq(b) ? bigInt(1) : bigInt(0);
    }

    gt(a, b) {
        return a.gt(b) ? bigInt(1) : bigInt(0);
    }

    leq(a, b) {
        return a.leq(b) ? bigInt(1) : bigInt(0);
    }

    geq(a, b) {
        return a.geq(b) ? bigInt(1) : bigInt(0);
    }

    neq(a, b) {
        return a.neq(b) ? bigInt(1) : bigInt(0);
    }

    div(a, b) {
        assert(!b.isZero(), "Division by zero");
        return a.mul(b.modInv(this.p)).mod(this.p);
    }

    idiv(a, b) {
        assert(!b.isZero(), "Division by zero");
        return a.divide(b);
    }

    mod(a, b) {
        return a.mod(b);
    }

    pow(a, b) {
        return a.modPow(b, this.p);
    }

    band(a, b) {
        return a.and(b).and(this.mask);
    }

    bor(a, b) {
        return a.or(b).and(this.mask);
    }

    bxor(a, b) {
        return a.xor(b).and(this.mask);
    }

    bnot(a) {
        return a.xor(this.mask).and(this.mask);
    }

    shl(a, b) {
        if (b.geq(this.bitLength)) return bigInt.zero;
        return a.shiftLeft(b).and(this.mask);
    }

    shr(a, b) {
        if (b.geq(this.bitLength)) return bigInt.zero;
        return a.shiftRight(b).and(this.mask);
    }

    land(a, b) {
        return (a.isZero() || b.isZero) ? bigInt.zero : bigInt.one;
    }

    lor(a, b) {
        return (a.isZero() && b.isZero) ? bigInt.zero : bigInt.one;
    }

    lnot(a) {
        return a.isZero() ? bigInt.one : bigInt.zero;
    }
};

